src/coreneuron/io/nrn_setup.cpp

```diff
@@ -207,21 +207,18 @@ void nrn_read_filesdat(int& ngrp, int*& grp, const char* filesdat) {
             "Info : The number of input datasets are less than ranks, some ranks will be idle!\n");
     }
 
-    ngrp = 0;
-    grp = new int[iNumFiles / nrnmpi_numprocs + 1];
```

Comment:
Yes. We no longer know a-priori how big the container will be. We will allocate multiple times. Do you think that is a problem?

Reply:
If you find critical we can reserve() the same amount of elements as before. Worse case we waste some memory, but we get the exact same behavior as before

Reply:
The key here is the comment in coreneuron/io/nrn_setup.cpp.
```
    // temporary bug work around. If any process has multiple threads, no
    // process can have a single thread. So, for now, if one thread, make two.
    // Fortunately, empty threads work fine.
    // Allocate NrnThread* nrn_threads of size ngroup (minimum 2)
    // Note that rank with 0 dataset/cellgroup works fine
```
If this issue no longer exists for MPI ranks where some ranks have 1 thread and other ranks have more than one thread, (or I never knew what I was talking about to begin with) then the work around can be abandoned.

Reply:
Ok, thanks for confirming. I kept that +1 behavior to be on the safe side. If anyone confirms we no longer need it I'm happy to drop.

In this thread I think Luc was referring to the several reallocations that will occur as the vector grows. I think this is of little impact since AFAIK each rank will normally have a few groups only - the same as the number of cycles. Are there exceptions to this?

Reply:
> normally have a few groups only ... Are there exceptions to this?

I don't know what a neurodamus cycle is, but a group is a subset of cells which become the contents of a pthread.
The number of groups on an MPI rank is likely to be on the order of the number of cores devoted to that rank.

At present, for direct transfer mode, coreneuron ranks/threads identically reflect neuron ranks/threads.
For file transfer mode, it is presumed that the groups can independently be assigned a rank/thread that is appropriate for load balance. (this load balance must be performed by ordering the files.dat contents)

Reply:
Ok, thanks. I guess it means that in some cases people might use more groups and less ranks. I can add the pre-allocation, it will have some benefit and in the general case it will be basically for free.